File: src/analysis.rs
use serde::{Deserialize, Serialize};

use crate::game::{evaluate, parse_code, Code, Feedback, Game, GameStatus, NUM_COLORS};

/// All possible codes in the classic 4-peg, 6-color game.
fn all_codes() -> Vec<Code> {
    let mut codes = Vec::with_capacity(1296); // 6^4
    for a in 1..=NUM_COLORS {
        for b in 1..=NUM_COLORS {
            for c in 1..=NUM_COLORS {
                for d in 1..=NUM_COLORS {
                    codes.push(parse_code(&[a, b, c, d]).unwrap());
                }
            }
        }
    }
    codes
}

/// Analysis result for a single turn
#[derive(Debug, Serialize, Deserialize)]
pub struct TurnAnalysis {
    pub attempt: usize,
    pub guess: Vec<u8>,
    pub feedback: Feedback,
    /// Number of remaining possible secrets before this guess
    pub candidates_before: usize,
    /// Number of remaining possible secrets after this guess
    pub candidates_after: usize,
    /// A suggested optimal guess (minimax) if the player's guess was suboptimal
    pub suggested_guess: Option<Vec<u8>>,
    /// The minimax worst-case remaining candidates for the suggested guess
    pub suggested_worst_case: Option<usize>,
    /// The worst-case remaining candidates for the player's actual guess
    pub actual_worst_case: usize,
    /// True when the player's guess was an optimal move
    pub was_optimal: bool,
}

/// Overall game analysis
#[derive(Debug, Serialize, Deserialize)]
pub struct GameAnalysis {
    pub game_id: String,
    pub status: GameStatus,
    pub total_attempts: usize,
    pub max_attempts: usize,
    pub secret: Vec<u8>,
    pub turns: Vec<TurnAnalysis>,
    /// A rating 0–100 (100 = perfect play)
    pub optimality_score: u8,
    /// Human-readable summary
    pub summary: String,
}

/// Computes the minimax score (worst-case partition size) for a given guess
/// against a set of remaining candidates.
fn minimax_score(guess: Code, candidates: &[Code]) -> usize {
    let mut partition: std::collections::HashMap<(u8, u8), usize> =
        std::collections::HashMap::new();
    for &secret in candidates {
        let fb = evaluate(secret, guess);
        *partition.entry((fb.blacks, fb.whites)).or_insert(0) += 1;
    }
    partition.values().copied().max().unwrap_or(0)
}

/// Finds the best guess from all 1296 codes using the minimax strategy,
/// preferring candidates over non-candidates on ties.
fn best_minimax_guess(candidates: &[Code]) -> (Code, usize) {
    if candidates.len() == 1 {
        return (candidates[0], 1);
    }

    let all = all_codes();
    let candidate_set: std::collections::HashSet<_> = candidates.iter().copied().collect();

    let mut best_guess = all[0];
    let mut best_score = usize::MAX;
    let mut best_is_candidate = false;

    for &g in &all {
        let score = minimax_score(g, candidates);
        let is_candidate = candidate_set.contains(&g);

        let better = score < best_score
            || (score == best_score && is_candidate && !best_is_candidate);

        if better {
            best_score = score;
            best_guess = g;
            best_is_candidate = is_candidate;
        }
    }

    (best_guess, best_score)
}

/// Filters a list of candidate codes to those consistent with the feedback
/// received for a particular guess.
fn filter_candidates(candidates: &[Code], guess: Code, feedback: Feedback) -> Vec<Code> {
    candidates
        .iter()
        .copied()
        .filter(|&candidate| evaluate(candidate, guess) == feedback)
        .collect()
}

/// Analyzes a completed (won or lost) game.
pub fn analyze_game(game: &Game) -> GameAnalysis {
    let all = all_codes();
    let mut candidates: Vec<Code> = all.clone();

    let mut turn_analyses: Vec<TurnAnalysis> = Vec::new();
    let mut suboptimal_turns: usize = 0;

    for turn in &game.turns {
        let candidates_before = candidates.len();
        let guess = parse_code(&turn.guess).expect("stored guess is always valid");
        let feedback = turn.feedback;

        let actual_worst_case = minimax_score(guess, &candidates);
        let (best_guess, best_score) = best_minimax_guess(&candidates);

        let was_optimal = actual_worst_case <= best_score;

        let suggested_guess = if was_optimal {
            None
        } else {
            suboptimal_turns += 1;
            Some(best_guess.map(|p| p.0).to_vec())
        };

        candidates = filter_candidates(&candidates, guess, feedback);
        let candidates_after = candidates.len();

        turn_analyses.push(TurnAnalysis {
            attempt: turn.attempt,
            guess: turn.guess.clone(),
            feedback,
            candidates_before,
            candidates_after,
            suggested_guess,
            suggested_worst_case: if was_optimal { None } else { Some(best_score) },
            actual_worst_case,
            was_optimal,
        });
    }

    let total_attempts = game.turns.len();
    let optimality_score = if total_attempts == 0 {
        100
    } else {
        let optimal_turns = total_attempts - suboptimal_turns;
        ((optimal_turns * 100) / total_attempts) as u8
    };

    let summary = build_summary(game, &turn_analyses, optimality_score);

    GameAnalysis {
        game_id: game.id.clone(),
        status: game.status.clone(),
        total_attempts,
        max_attempts: crate::game::MAX_ATTEMPTS,
        secret: game.secret.map(|p| p.0).to_vec(),
        turns: turn_analyses,
        optimality_score,
        summary,
    }
}

fn build_summary(game: &Game, turns: &[TurnAnalysis], score: u8) -> String {
    match &game.status {
        GameStatus::InProgress => "Game is still in progress.".to_string(),
        GameStatus::Won => {
            let suboptimal: Vec<_> = turns.iter().filter(|t| !t.was_optimal).collect();
            if suboptimal.is_empty() {
                format!(
                    "Excellent! You solved the puzzle in {} attempt(s) with optimal play (score: {}/100).",
                    game.turns.len(),
                    score
                )
            } else {
                format!(
                    "You solved the puzzle in {} attempt(s) with an optimality score of {}/100. \
                     {} turn(s) had a suboptimal guess — check the per-turn suggestions.",
                    game.turns.len(),
                    score,
                    suboptimal.len()
                )
            }
        }
        GameStatus::Lost => {
            format!(
                "You ran out of guesses after {} attempt(s) (score: {}/100). \
                 The secret was {:?}. Review the suggestions to improve your strategy.",
                game.turns.len(),
                score,
                game.secret.map(|p| p.0)
            )
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::game::{Game, Peg};

    fn code(a: u8, b: u8, c: u8, d: u8) -> Code {
        [Peg(a), Peg(b), Peg(c), Peg(d)]
    }

    #[test]
    fn test_all_codes_count() {
        assert_eq!(all_codes().len(), 1296);
    }

    #[test]
    fn test_filter_candidates() {
        let all = all_codes();
        let guess = code(1, 1, 2, 2);
        // If we get 2 blacks, 0 whites -> many consistent codes
        let fb = Feedback { blacks: 2, whites: 0 };
        let filtered = filter_candidates(&all, guess, fb);
        // Every remaining code should produce exactly that feedback against guess
        for c in &filtered {
            assert_eq!(evaluate(*c, guess), fb);
        }
    }

    #[test]
    fn test_analyze_won_game_optimal() {
        // Use Knuth's known optimal first guess 1122
        let mut game = Game::new("test".into());
        game.secret = code(1, 2, 3, 4);

        // Simulate known-good guesses
        game.guess(code(1, 1, 2, 2)).unwrap(); // well-known optimal first guess
        game.guess(code(1, 2, 3, 4)).unwrap(); // win

        let analysis = analyze_game(&game);
        assert_eq!(analysis.status, GameStatus::Won);
        assert_eq!(analysis.total_attempts, 2);
        assert!(analysis.optimality_score > 0);
    }

    #[test]
    fn test_analyze_lost_game() {
        let mut game = Game::new("test".into());
        game.secret = code(1, 2, 3, 4);

        for _ in 0..crate::game::MAX_ATTEMPTS {
            let _ = game.guess(code(5, 5, 5, 5));
        }

        let analysis = analyze_game(&game);
        assert_eq!(analysis.status, GameStatus::Lost);
        assert_eq!(analysis.secret, vec![1, 2, 3, 4]);
    }
}

File: mastermind-go/internal/analysis/analysis.go
// Package analysis implements minimax-based game analysis for Mastermind.
package analysis

import (
	"fmt"

	"github.com/j4bberwocky/mastermind/mastermind-go/internal/game"
)

// TurnAnalysis holds the analysis result for a single turn.
type TurnAnalysis struct {
	Attempt            int           `json:"attempt"`
	Guess              []uint8       `json:"guess"`
	Feedback           game.Feedback `json:"feedback"`
	CandidatesBefore   int           `json:"candidates_before"`
	CandidatesAfter    int           `json:"candidates_after"`
	SuggestedGuess     []uint8       `json:"suggested_guess"`
	SuggestedWorstCase *int          `json:"suggested_worst_case"`
	ActualWorstCase    int           `json:"actual_worst_case"`
	WasOptimal         bool          `json:"was_optimal"`
}

// GameAnalysis holds the overall game analysis.
type GameAnalysis struct {
	GameID          string          `json:"game_id"`
	Status          game.GameStatus `json:"status"`
	TotalAttempts   int             `json:"total_attempts"`
	MaxAttempts     int             `json:"max_attempts"`
	Secret          []uint8         `json:"secret"`
	Turns           []TurnAnalysis  `json:"turns"`
	OptimalityScore uint8           `json:"optimality_score"`
	Summary         string          `json:"summary"`
}

// AllCodes generates all possible codes in the classic 4-peg, 6-color game.
func AllCodes() []game.Code {
	codes := make([]game.Code, 0, 1296) // 6^4
	for a := uint8(1); a <= game.NumColors; a++ {
		for b := uint8(1); b <= game.NumColors; b++ {
			for c := uint8(1); c <= game.NumColors; c++ {
				for d := uint8(1); d <= game.NumColors; d++ {
					codes = append(codes, game.Code{game.Peg(a), game.Peg(b), game.Peg(c), game.Peg(d)})
				}
			}
		}
	}
	return codes
}

// feedbackKey creates a key for a feedback pair to use in maps.
type feedbackKey struct {
	blacks, whites uint8
}

// MinimaxScore computes the worst-case partition size for a given guess
// against a set of remaining candidates.
func MinimaxScore(guess game.Code, candidates []game.Code) int {
	partition := make(map[feedbackKey]int)
	for _, secret := range candidates {
		fb := game.Evaluate(secret, guess)
		partition[feedbackKey{fb.Blacks, fb.Whites}]++
	}
	maxVal := 0
	for _, count := range partition {
		if count > maxVal {
			maxVal = count
		}
	}
	return maxVal
}

// BestMinimaxGuess finds the best guess from all 1296 codes using the minimax strategy,
// preferring candidates over non-candidates on ties.
func BestMinimaxGuess(candidates []game.Code) (game.Code, int) {
	if len(candidates) == 1 {
		return candidates[0], 1
	}

	all := AllCodes()
	candidateSet := make(map[game.Code]bool, len(candidates))
	for _, c := range candidates {
		candidateSet[c] = true
	}

	bestGuess := all[0]
	bestScore := int(^uint(0) >> 1) // max int
	bestIsCandidate := false

	for _, g := range all {
		score := MinimaxScore(g, candidates)
		isCandidate := candidateSet[g]

		better := score < bestScore ||
			(score == bestScore && isCandidate && !bestIsCandidate)

		if better {
			bestScore = score
			bestGuess = g
			bestIsCandidate = isCandidate
		}
	}

	return bestGuess, bestScore
}

// FilterCandidates filters a list of candidate codes to those consistent with the
// feedback received for a particular guess.
func FilterCandidates(candidates []game.Code, guess game.Code, feedback game.Feedback) []game.Code {
	var result []game.Code
	for _, candidate := range candidates {
		if game.Evaluate(candidate, guess) == feedback {
			result = append(result, candidate)
		}
	}
	return result
}

// AnalyzeGame analyzes a completed (won or lost) game.
func AnalyzeGame(g *game.Game) *GameAnalysis {
	all := AllCodes()
	candidates := make([]game.Code, len(all))
	copy(candidates, all)

	turnAnalyses := make([]TurnAnalysis, 0, len(g.Turns))
	suboptimalTurns := 0

	for _, turn := range g.Turns {
		candidatesBefore := len(candidates)
		guess, _ := game.ParseCode(turn.Guess)
		feedback := turn.Feedback

		actualWorstCase := MinimaxScore(guess, candidates)
		bestGuess, bestScore := BestMinimaxGuess(candidates)

		wasOptimal := actualWorstCase <= bestScore

		var suggestedGuess []uint8
		var suggestedWorstCase *int
		if !wasOptimal {
			suboptimalTurns++
			suggestedGuess = game.CodeToSlice(bestGuess)
			suggestedWorstCase = &bestScore
		}

		candidates = FilterCandidates(candidates, guess, feedback)
		candidatesAfter := len(candidates)

		turnAnalyses = append(turnAnalyses, TurnAnalysis{
			Attempt:            turn.Attempt,
			Guess:              turn.Guess,
			Feedback:           feedback,
			CandidatesBefore:   candidatesBefore,
			CandidatesAfter:    candidatesAfter,
			SuggestedGuess:     suggestedGuess,
			SuggestedWorstCase: suggestedWorstCase,
			ActualWorstCase:    actualWorstCase,
			WasOptimal:         wasOptimal,
		})
	}

	totalAttempts := len(g.Turns)
	var optimalityScore uint8
	if totalAttempts == 0 {
		optimalityScore = 100
	} else {
		optimalTurns := totalAttempts - suboptimalTurns
		optimalityScore = uint8((optimalTurns * 100) / totalAttempts)
	}

	summary := buildSummary(g, turnAnalyses, optimalityScore)

	return &GameAnalysis{
		GameID:          g.ID,
		Status:          g.Status,
		TotalAttempts:   totalAttempts,
		MaxAttempts:     game.MaxAttempts,
		Secret:          game.CodeToSlice(g.Secret),
		Turns:           turnAnalyses,
		OptimalityScore: optimalityScore,
		Summary:         summary,
	}
}

func buildSummary(g *game.Game, turns []TurnAnalysis, score uint8) string {
	switch g.Status {
	case game.StatusInProgress:
		return "Game is still in progress."
	case game.StatusWon:
		suboptimal := 0
		for _, t := range turns {
			if !t.WasOptimal {
				suboptimal++
			}
		}
		if suboptimal == 0 {
			return fmt.Sprintf(
				"Excellent! You solved the puzzle in %d attempt(s) with optimal play (score: %d/100).",
				len(g.Turns), score,
			)
		}
		return fmt.Sprintf(
			"You solved the puzzle in %d attempt(s) with an optimality score of %d/100. "+
				"%d turn(s) had a suboptimal guess — check the per-turn suggestions.",
			len(g.Turns), score, suboptimal,
		)
	case game.StatusLost:
		secret := game.CodeToSlice(g.Secret)
		return fmt.Sprintf(
			"You ran out of guesses after %d attempt(s) (score: %d/100). "+
				"The secret was %v. Review the suggestions to improve your strategy.",
			len(g.Turns), score, secret,
		)
	default:
		return "Unknown game status."
	}
}
